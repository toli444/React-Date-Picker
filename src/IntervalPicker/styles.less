.interval-picker {
  margin-top: 30vh;
}
