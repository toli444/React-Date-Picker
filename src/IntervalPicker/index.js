import React, {Component} from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

import styles from './styles.less';

import DatePicker from './components/DatePicker';

class IntervalPicker extends Component {
  static propTypes = {
    startDate: PropTypes.object,
    endDate: PropTypes.object,
    onStartDateChange: PropTypes.func,
    onEndDateChange: PropTypes.func,
    minDate: PropTypes.object,
    maxDate: PropTypes.object,
    format: PropTypes.string,
  };

  static defaultProps = {
    startDate: moment(),
    endDate: moment().add(3, 'month'),
    maxDate: moment().add(3, 'years'),
    minDate: moment().subtract(3, 'years'),
    format: 'DD-MM-YYYY',
    onStartDateChange: () => {},
    onEndDateChange: () => {},
  };

  constructor(props) {
    super(props);

    this.state = {
      startDate: props.startDate ? moment(props.startDate, props.format) : null,
      endDate: props.endDate ? moment(props.endDate, props.format) : null,
    };

    this.endDatePickerRef = React.createRef();
  }

  componentDidMount() {
    this.props.onStartDateChange(this.props.startDate);
    this.props.onEndDateChange(this.props.endDate);
  }

  handleStartDateChange = value => {
    this.setState({startDate: value});
    this.props.onStartDateChange(value);

    if (value.isSameOrAfter(this.state.endDate)) {
      this.setState({endDate: null}, () => {
        this.endDatePickerRef.current.focusIn();
      });
    } else {
      this.endDatePickerRef.current.focusIn();
    }
  };

  handleEndDateChange = value => {
    this.setState({endDate: value});
    this.props.onEndDateChange(value);
  };

  render() {
    const {minDate, maxDate, format} = this.props;
    const {startDate, endDate} = this.state;

    return (
      <div className={styles['interval-picker']}>
        <DatePicker
          value={startDate}
          startDate={startDate}
          endDate={endDate}
          onDateChange={value => this.handleStartDateChange(value)}
          minDate={minDate}
          maxDate={maxDate}
          format={format}
          label={'Date from:'}
        />
        <DatePicker
          className="right"
          value={endDate}
          startDate={startDate}
          endDate={endDate}
          onDateChange={value => this.handleEndDateChange(value)}
          minDate={startDate}
          maxDate={maxDate}
          format={format}
          ref={this.endDatePickerRef}
          label={'Date to:'}
        />
      </div>
    );
  }
}

export default IntervalPicker;
