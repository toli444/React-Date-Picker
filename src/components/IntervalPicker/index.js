import React, {Component} from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

import styles from './styles.less';

import DatePicker from './components/DatePicker/index';

class IntervalPicker extends Component {
  static propTypes = {
    startDate: PropTypes.object,
    endDate: PropTypes.object,
    onStartDateChange: PropTypes.func,
    onEndDateChange: PropTypes.func,
    minDate: PropTypes.object,
    maxDate: PropTypes.object,
    format: PropTypes.string,
  };

  static defaultProps = {
    startDate: moment(),
    endDate: moment(),
    maxDate: moment().add(3, 'years'),
    minDate: moment().subtract(3, 'years'),
    format: 'DD-MM-YYYY',
    onStartDateChange: () => {},
    onEndDateChange: () => {},
  };

  constructor(props) {
    super(props);

    this.endDatePickerRef = React.createRef();
  }

  handleStartDateChange = value => {
    const {endDate, onStartDateChange, onEndDateChange} = this.props;
    onStartDateChange(value);

    if (value.isSameOrAfter(endDate)) {
      onEndDateChange(value);
    }

    setImmediate(() => {
      this.endDatePickerRef.current.focusIn();
    });
  };

  render() {
    const {minDate, maxDate, format, onEndDateChange, startDate, endDate} = this.props;

    return (
      <div className={styles['interval-picker']}>
        <DatePicker
          value={startDate}
          startDate={startDate}
          endDate={endDate}
          onDateChange={value => this.handleStartDateChange(value)}
          minDate={minDate}
          maxDate={maxDate}
          format={format}
          label={'Date from:'}
        />
        <DatePicker
          className="right"
          value={endDate}
          startDate={startDate}
          endDate={endDate}
          onDateChange={onEndDateChange}
          minDate={startDate}
          maxDate={maxDate}
          format={format}
          ref={this.endDatePickerRef}
          label={'Date to:'}
        />
      </div>
    );
  }
}

export default IntervalPicker;
