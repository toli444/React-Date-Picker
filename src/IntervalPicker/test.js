import React from 'react';
import { shallow } from 'enzyme';
import IntervalPicker from '../IntervalPicker';
import moment from "moment";

const format = 'DD/MM/YYYY';

describe('IntervalPicker', () => {
  const RealDate = Date;

  const mockDate = isoDate => {
    global.Date = class extends RealDate {
      constructor () {
        super();

        return new RealDate(isoDate)
      }
    }
  };

  afterEach(() => {
    jest.clearAllMocks();
    global.Date = RealDate
  });

  it('renders correctly', () => {
    mockDate('2017-11-25T12:34:56z');
    const wrapper  = shallow(<IntervalPicker />);

    expect(wrapper).toMatchSnapshot();
  });

  it('componentDidMount', () => {
    const onStartDateChangeStub = jest.fn();
    const onEndDateChangeStub = jest.fn();
    const startDate = moment('28/07/1997', format);
    const endDate = moment('29/07/1997', format);
    const wrapper  = shallow(<IntervalPicker
      onStartDateChange={onStartDateChangeStub}
      onEndDateChange={onEndDateChangeStub}
      startDate={startDate}
      endDate={endDate}
    />);

    wrapper.instance().componentDidMount();
    expect(onStartDateChangeStub).toBeCalledWith(startDate);
    expect(onEndDateChangeStub).toBeCalledWith(endDate);
  });

  it('handleStartDateChange', () => {
    const onStartDateChangeStub = jest.fn();
    const wrapper  = shallow(<IntervalPicker
      onStartDateChange={onStartDateChangeStub}
    />);
    const instance = wrapper.instance();
    instance.endDatePickerRef.current = { focusIn: jest.fn() };

    const execTestCase = (startDate, endDate, expectedEndDate = endDate) => {
      instance.state.endDate = endDate;
      instance.handleStartDateChange(startDate);
      expect(onStartDateChangeStub).toBeCalledWith(startDate);
      expect(instance.state.endDate).toBe(expectedEndDate);
      expect(instance.endDatePickerRef.current.focusIn).toBeCalled();
      jest.clearAllMocks();
    };

    execTestCase(moment('28/07/1997', format), moment('29/07/1997', format));
    execTestCase(moment('29/07/1997', format), moment('29/07/1997', format), null);
    execTestCase(moment('30/07/1997', format), moment('29/07/1997', format), null);
  });

  it('handleEndDateChange', () => {
    const onEndDateChangeStub = jest.fn();
    const wrapper  = shallow(<IntervalPicker
      onEndDateChange={onEndDateChangeStub}
    />);
    const instance = wrapper.instance();

    const execTestCase = (newEndDate, expectedStateEndDate = newEndDate) => {
      instance.handleEndDateChange(newEndDate);
      expect(instance.state.endDate).toBe(expectedStateEndDate);
      expect(onEndDateChangeStub).toBeCalledWith(newEndDate);
      jest.clearAllMocks();
    };

    execTestCase(moment('28/07/1997', format));
    execTestCase(moment('29/07/1997', format));
    execTestCase(moment('30/07/1997', format));
  });
});
