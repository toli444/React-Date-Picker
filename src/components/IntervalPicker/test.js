import React from 'react';
import { shallow } from 'enzyme';
import IntervalPicker from './index';
import moment from "moment";

const format = 'DD/MM/YYYY';

describe('IntervalPicker', () => {
  const RealDate = Date;

  const mockDate = isoDate => {
    global.Date = class extends RealDate {
      constructor () {
        super();

        return new RealDate(isoDate)
      }
    }
  };

  afterEach(() => {
    jest.clearAllMocks();
    global.Date = RealDate
  });

  it('renders correctly', () => {
    mockDate('2017-11-25T12:34:56z');
    const wrapper  = shallow(<IntervalPicker />);

    expect(wrapper).toMatchSnapshot();
  });

  it('handleStartDateChange', () => {
    const onStartDateChangeStub = jest.fn();
    const onEndDateChangeStub = jest.fn();
    const wrapper  = shallow(<IntervalPicker
      onStartDateChange={onStartDateChangeStub}
      onEndDateChange={onEndDateChangeStub}
    />);
    const instance = wrapper.instance();
    instance.endDatePickerRef.current = { focusIn: jest.fn() };

    const execTestCase = (startDate, endDate, endDateShouldBeCalled) => {
      wrapper.setProps({ endDate });
      instance.handleStartDateChange(startDate);
      expect(onStartDateChangeStub).toBeCalledWith(startDate);
      endDateShouldBeCalled && expect(onEndDateChangeStub).toBeCalledWith(startDate);
      setImmediate(() => {expect(instance.endDatePickerRef.current.focusIn).toBeCalled()});
      jest.clearAllMocks();
    };

    execTestCase(moment('28/07/1997', format), moment('29/07/1997', format));
    execTestCase(moment('29/07/1997', format), moment('29/07/1997', format), true);
    execTestCase(moment('30/07/1997', format), moment('29/07/1997', format), true);
  });
});
