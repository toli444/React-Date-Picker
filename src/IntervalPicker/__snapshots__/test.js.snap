// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntervalPicker renders correctly 1`] = `
<div
  className="interval-picker"
>
  <DatePicker
    className=""
    endDate={"2017-11-25T12:34:56.000Z"}
    format="DD-MM-YYYY"
    label="Date from:"
    maxDate={"2017-11-25T12:34:56.000Z"}
    minDate={"2017-11-25T12:34:56.000Z"}
    onDateChange={[Function]}
    startDate={"2017-11-25T12:34:56.000Z"}
    value={"2017-11-25T12:34:56.000Z"}
  />
  <DatePicker
    className="right"
    endDate={"2017-11-25T12:34:56.000Z"}
    format="DD-MM-YYYY"
    label="Date to:"
    maxDate={"2017-11-25T12:34:56.000Z"}
    minDate={"2017-11-25T12:34:56.000Z"}
    onDateChange={[Function]}
    startDate={"2017-11-25T12:34:56.000Z"}
    value={"2017-11-25T12:34:56.000Z"}
  />
</div>
`;
